src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs

```diff
@@ -259,19 +314,60 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
                     }
 
                     if (hasAlias == false)
-                        rewriteForProjectionFromMapReduceIndex = false;
+                        queryChanges &= ~QueryChanges.RewriteForProjectionFromMapReduceIndex;
                 }
             }
         }
 
-        if (rewriteForPaging == false && rewriteForProjectionFromMapReduceIndex == false && rewriteForProjectionFromAutoMapReduceIndex == false)
+        if (queryChanges == QueryChanges.None)
             return queryTemplate;
 
         var clone = Query.Metadata.Query.ShallowCopy();
 
         DynamicJsonValue modifications = new(queryTemplate);
 
-        if (rewriteForPaging)
+        if (queryChanges.HasFlag(QueryChanges.RewriteForFilterInMapReduce))
+        {
+            clone.Filter = null;
+            clone.FilterLimit = null;
+
+            RemoveLimitAndOffset();
+        }
+
+        if (queryChanges.HasFlag(QueryChanges.RewriteForLimitWithOrderByInMapReduce))
+        {
+            // we have a limit with order by
+            clone.OrderBy = null;
+
+            RemoveLimitAndOffset();
+        }
+
+        if (queryChanges.HasFlag(QueryChanges.UpdateOrderByFieldsInMapReduce))
+        {
+            clone.OrderBy ??= new List<(QueryExpression Expression, OrderByFieldType FieldType, bool Ascending)>();
+            OrderByFields = new List<OrderByField>();
+
+            if (Query.Metadata.OrderBy != null)
+            {
+                // add existing order by fields
+                foreach (var orderByField in Query.Metadata.OrderBy)
+                {
+                    OrderByFields.Add(orderByField);
+                }
+            }
+
+            // add the missing reduce keys
+            foreach (var reduceKey in reduceKeys)
```

Comment:
I think we should guarantee the order of  added group by fields. `HashSet` doesn't guarantee that

Reply:
done